internal/providers/factory.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2024 Stacklok, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package providers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	serverconfig "github.com/stacklok/minder/internal/config/server"
+	"github.com/stacklok/minder/internal/crypto"
+	"github.com/stacklok/minder/internal/db"
+	gitclient "github.com/stacklok/minder/internal/providers/git"
+	githubapp "github.com/stacklok/minder/internal/providers/github/app"
+	ghclient "github.com/stacklok/minder/internal/providers/github/oauth"
+	httpclient "github.com/stacklok/minder/internal/providers/http"
+	"github.com/stacklok/minder/internal/providers/ratecache"
+	"github.com/stacklok/minder/internal/providers/telemetry"
+	provinfv1 "github.com/stacklok/minder/pkg/providers/v1"
+)
+
+// TraitInstantiator is responsible for creating instances of the Profile
+// interfaces defined in `github.com/stacklok/minder/pkg/providers/v1` from
+// the providers stored in the database.
+type TraitInstantiator interface {
+	GetGit(ctx context.Context, provider *db.Provider) (provinfv1.Git, error)
+	GetHTTP(ctx context.Context, provider *db.Provider) (provinfv1.REST, error)
+	GetGitHub(ctx context.Context, provider *db.Provider, options *GitHubOptions) (provinfv1.GitHub, error)
+	GetRepoLister(ctx context.Context, provider *db.Provider) (provinfv1.RepoLister, error)
+}
+
+// GitHubOptions contains options used when creating a GitHub instance
+// If omitted, these will be looked up in the database
+type GitHubOptions struct {
+	Credential  provinfv1.GitHubCredential
```

Comment:
We'll use credentials in all providers, why is this specific to GitHub? Did you check the OCI Providers POC?

Reply:
In most cases, the credentials are loaded from the database. However, there are a few places where the Github trait is instantiated with credentials which are not in the DB - one is in `internal/providers/service.go` (which is in this PR), and there are two in the CLI (not in this PR). In those cases, it is necessary to provide some Github specific details, namely the owner filter and the credentials used.

Your comment did make me realize that there was a bug in my change where I assumed Github credentials in more places than I should have. I fixed that in the most recent commit.